src/ansys/grantami/recordlists/_connection.py

```diff
@@ -530,6 +545,62 @@ def unsubscribe_from_list(self, record_list: RecordList) -> None:
         )
 
 
+class _ItemResolver:
+    _max_requests = 5
+
+    def __init__(self, client: ApiClient) -> None:
+        self._record_histories_api = api.RecordsRecordHistoriesApi(client)
+        self._db_schema_api = api.SchemaDatabasesApi(client)
+        self._db_map: Dict[str, str] = {}
+
+    def get_resolvable_items(self, all_items: List[RecordListItem]) -> List[RecordListItem]:
+        """Test all items to see if they can be resolved as the current user.
+
+        Uses concurrent.futures to handle threading, since openapi-common is currently based on
+        requests which limits asyncio options.
+
+        Parameters
+        ----------
+        all_items
+            The items for which to check resolvability.
+
+        Returns
+        -------
+        resolvable_items
+            The items which could be resolved on the server.
+        """
+        self._db_map = self._get_db_map()
+        with concurrent.futures.ThreadPoolExecutor(max_workers=self._max_requests) as executor:
+            resolvable_test_futures = {
+                executor.submit(self._is_item_resolvable, i): i for i in all_items
+            }
+            resolvable_items = [
+                resolvable_test_futures[f]
+                for f in concurrent.futures.as_completed(resolvable_test_futures)
+                if f.result()
+            ]
+        return resolvable_items
+
+    def _get_db_map(self) -> Dict[str, str]:
+        dbs = self._db_schema_api.get_all_databases()
+        return {db.guid: db.key for db in dbs.databases}
+
+    def _is_item_resolvable(self, item: RecordListItem) -> bool:
+        if item.database_guid not in self._db_map:
+            return False
+        try:
+            self._record_histories_api.get_record_history(
```

Comment:
We might instead need to test `get_record_version` if the `RecordListItem` has version information.

@ludovicsteinbach is the behavior that if the `RecordListItem` doesn't have version information, it tracks the latest history, but if it does then it refers to a specific version?

If that's the case, I think the one scenario we'd currently miss is if an unreleased version of a VC record is specifically added to a list, and then a read user attempts to read it. This check will pass if there is a currently-released V1 or greater of that record, but attempting to resolve the specific version guid will fail.

Reply:
I think the behavior should be:

* If the `RecordListItem` doesn't define a version, check the user can resolve the history. This will fail if
  * The history has been deleted
  * The history has been withdrawn and the user is a read user
  * The history hasn't yet been released and the user is a read user
  * The history is hidden by access control
* If the `RecordListItem` defines a version, check the user can resolve the specific version. This will fail if 
  * The version has not yet been released, 
  * The version been withdrawn and the user is a read user, 
  * The history is hidden by access control

The intention of this change is to take into account the current user's permissions when assessing if they can resolve a record. So two different users may see two different lists of `RecordListItem`s. Maybe the new method should clarify this explicitly. It could also accept a parameter that gets passed into the `x_ansys_vc_mode` that forces a read-only view of the database.

Reply:
So this kind of snowballed. I've added fixtures that set up the records we need in the state that's required. Since it's version controlled we can't undo these changes, so I've written it to first check if a record exists in the appropriate state.

Added all sorts of permutations of checks to see how different version states are handled.